src/Chapter4.hs

```diff
@@ -488,7 +495,17 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
+instance Applicative List where
+  pure :: a -> List a
+  pure a = Cons a Empty
 
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty <*> _ = Empty
+  (Cons f fs) <*> x = concatList (fmap f x) (fs <*> x)
+
+concatList :: List a -> List a -> List a
+concatList Empty x = x
+concatList (Cons a as) x = Cons a (concatList as x)
```

Comment:
Perfect!

You can one more case to make your solution even more efficient:

```haskell
concatList xs Empty = xs
```

In that case, you won't need to walk through each element of the first list when the second one is empty and return the list instantly 🚤